docs/proposals/cluster-capacity.md

```diff
@@ -0,0 +1,461 @@
+* Provide REST-API to query questions like "approximately how many more pods could i schedule with this shape?"
+* Integrate the cluster capacity into ``kubectl`` (e.g. ``kubectl cluster-capacity``)
+
+### Code analysis
+
+Currently, each iteration of the scheduler consists of the following [steps](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/scheduler.go#L93):
+
+1. ask for the next pod ``s.config.NextPod``
+1. schedule the pod ``s.config.Algorithm.Schedule``
+1. update metrics ``metrics.SchedulingAlgorithmLatency.Observe``
+1. update scheduler cache ``s.config.SchedulerCache.AssumePod``
+1. pod binding ``s.config.Binder.Bind``
+
+Scheduling itself (``Schedule``) consists of the following [steps](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/generic_scheduler.go#L79):
+
+1. retrieve a list of nodes ``nodeLister.List``
+1. update node info ``g.cache.UpdateNodeNameToInfoMap``
+1. filter nodes with predicates ``findNodesThatFit``
+1. prioritized remaining nodes ``PrioritizeNodes``
+1. pick the most suitable node ``g.selectHost``
+
+Scheduler keeps a number of caches which capture the current state of the cluster.
+Pod cache is continuosly updated [outside](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L128) the scheduler.
+The same holds for the [list of nodes](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L140).
+
+As the individual predicates and priority functions need to query other objects such as services, controllers, etc.,
+other caches are populated as well:
+
+* PV cache
+* PVC cache
+* Service cache
+* Controller cache
+* ReplicaSet cache
+
+All caches are continuosly updated via [reflectors](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L387):
+
+```
+cache.NewReflector(f.createPersistentVolumeLW(), &api.PersistentVolume{}, f.PVLister.Store, 0).RunUntil(f.StopEverything)
+cache.NewReflector(f.createPersistentVolumeClaimLW(), &api.PersistentVolumeClaim{}, f.PVCLister.Store, 0).RunUntil(f.StopEverything)
+cache.NewReflector(f.createServiceLW(), &api.Service{}, f.ServiceLister.Store, 0).RunUntil(f.StopEverything)
+cache.NewReflector(f.createControllerLW(), &api.ReplicationController{}, f.ControllerLister.Indexer, 0).RunUntil(f.StopEverything)
+cache.NewReflector(f.createReplicaSetLW(), &extensions.ReplicaSet{}, f.ReplicaSetLister.Store, 0).RunUntil(f.StopEverything)
+```
+
+Once the ``f.StopEverything`` channel is closed, no cache is updated anymore.
+
+Each iteration of the scheduler calls ``s.config.NextPod()`` function which pops one pod at a time from the queue of unscheduled pods.
+The queue is again continuously updated via [reflector](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L389).
+
+Once the scheduler decides what node to schedule a pod on, the pod is bind to the Apiserver.
```

Comment:
For a 1st iteration, this may be ok.  But we need to plan for eventual autonomy as well as history.